File: languages/julia/test/test_conformance.jl
using AgentBible
using JSON3
using Test

@testset "Conformance" begin
    hadamard = ComplexF64[1.0 1.0; 1.0 -1.0] / sqrt(2.0)
    @test AgentBible.check_unitary(hadamard) == hadamard

    perturbed = copy(hadamard)
    theta = sqrt(2.0) * 1e-9
    perturbed[:, 2] .*= exp(im * theta)
    @test AgentBible.check_unitary(perturbed) == perturbed

    failing = copy(hadamard)
    failing[1, 1] += 1e-5
    @test_throws AgentBibleError AgentBible.check_unitary(failing)

    @test_throws AgentBibleError AgentBible.check_finite(NaN)

    record = ProvenanceRecord([
        CheckResult("unitary", true, 1e-10, 1e-12, "frobenius", nothing),
    ])
    io = IOBuffer()
    emit_provenance(record, io)
    payload = JSON3.read(String(take!(io)))

    @test payload["spec_version"] == "1.0"
    @test payload["language"] == "julia"
    @test payload["checks_passed"][1]["check_name"] == "unitary"
end

File: languages/rust/agentbible/tests/test_conformance.rs
use std::panic::catch_unwind;

use ndarray::array;
use num_complex::Complex64;

use agentbible::provenance::{CheckResult, ProvenanceRecord};
use agentbible::types::UnitaryMatrix;
use agentbible::validate_finite;

fn hadamard() -> ndarray::Array2<Complex64> {
    array![
        [
            Complex64::new(1.0 / 2_f64.sqrt(), 0.0),
            Complex64::new(1.0 / 2_f64.sqrt(), 0.0),
        ],
        [
            Complex64::new(1.0 / 2_f64.sqrt(), 0.0),
            Complex64::new(-1.0 / 2_f64.sqrt(), 0.0),
        ]
    ]
}

#[test]
fn unitary_conformance_matches_spec_thresholds() {
    let base = hadamard();
    assert!(UnitaryMatrix::new(base.clone()).is_ok());

    let mut perturbed = base.clone();
    let theta = std::f64::consts::SQRT_2 * 1e-9;
    let phase = Complex64::new(0.0, theta).exp();
    perturbed[(0, 1)] *= phase;
    perturbed[(1, 1)] *= phase;
    assert!(UnitaryMatrix::new(perturbed).is_ok());

    let mut failing = base;
    failing[(0, 0)] += Complex64::new(1e-5, 0.0);
    assert!(UnitaryMatrix::new(failing).is_err());
}

#[test]
fn nan_injection_fails_finite() {
    assert!(catch_unwind(|| validate_finite!(f64::NAN)).is_err());
}

#[test]
fn provenance_json_roundtrip() {
    let record = ProvenanceRecord::new(vec![CheckResult::success(
        "unitary",
        1e-10,
        1e-12,
        "frobenius",
    )]);
    let encoded = record.to_json_string();
    let decoded: ProvenanceRecord =
        serde_json::from_str(&encoded).expect("provenance roundtrip should decode");
    assert_eq!(decoded.spec_version, "1.0");
    assert_eq!(decoded.language, "rust");
    assert_eq!(decoded.checks_passed[0].check_name, "unitary");
}
